blaze/expr/collections.py

```diff
@@ -435,35 +442,57 @@ def schema(self):
         """
         option = lambda dt: dt if isinstance(dt, Option) else Option(dt)
 
-        joined = [[name, dt] for name, dt in self.lhs.schema[0].parameters[0]
-                  if name in self.on_left]
+        on_left = self.on_left
+        if not isinstance(on_left, list):
+            on_left = on_left,
 
-        left = [[name, dt] for name, dt in
-                zip(self.lhs.fields, types_of_fields(
-                    self.lhs.fields, self.lhs))
-                if name not in self.on_left]
+        on_right = self.on_right
+        if not isinstance(on_right, list):
+            on_right = on_right,
 
-        right = [[name, dt] for name, dt in
-                 zip(self.rhs.fields, types_of_fields(
-                     self.rhs.fields, self.rhs))
-                 if name not in self.on_right]
+        right_types = keymap(
+            dict(zip(on_right, on_left)).get,
+            dict(self.rhs.schema[0].parameters[0]),
```

Comment:
just tried this, you can do `self.rhs.dshape.measure.dict` and remove the `dict()` call around it